feat: add ESO and ESC blog post (#12962)
pulumi/docs: content/blog/cloud-native-secret-management-with-pulumi-esc-and-external-secrets-operator/index.md

---
title: "Pulumi ESC and External Secrets Operator: The Perfect Solution for Today's Cloud-Native Secret Management"
allow_long_title: true
date: "2024-10-01"
meta_desc: "Learn how to manage Kubernetes secrets with Pulumi ESC and External Secrets Operator"
meta_image: meta.png
authors:
- engin-diri

tags:
- pulumi
- esc
- external-secrets-operator
- kubernetes
- cloud-native
- security
- secret-management
---

Managing secrets in a cloud-native environment can be challenging, but it is crucial for ensuring the security and integrity of any application or infrastructure. We encounter a lot of different types of secrets, from API keys, database passwords, and certificates to tokens and passwords. These secrets need to be stored securely and accessed by different services in a secure way without exposing any sensitive information to unauthorized users.

> Here is where Pulumi ESC and External Secrets Operator come into play by providing a secure and efficient solution for cloud-native secret management.

Reviewer: ```suggestion
Here is where [Pulumi ESC](/product/secrets-management/) and External Secrets Operator come into play by providing a secure and efficient solution for cloud-native secret management.
```

Reviewer: Also link to our ESO integrations page here

## Pulumi ESC (Environments, Secrets, and Configuration)

> Pulumi ESC is our new secrets management & orchestration service for environments, secrets, and configurations. It can be used as part of your Pulumi IaC (Infrastructure as Code) projects using `pulumiConfig` to expose the stored values to your Pulumi stacks.

Reviewer: I'd rearrange this to say "Pulumi ESC can do X, Y, Z for your applications because of A, B, C features... and Also btw, you can also use it for Infrastructure as code - both Pulumi IaC and Terraform"

## External Secrets Operator (ESO)

The Open Source project [External Secrets Operator](https://external-secrets.io/latest/) (ESO) is a Kubernetes operator that helps manage secrets in a highly secure and scalable way. ESO synchronizes secrets stored in external secret management systems like Pulumi ESC, HashiCorp Vault, AWS Secrets Manager, Azure Key Vault, and more into Kubernetes secrets. This approach allows us to control secret sprawl and centralize the management of secrets in a single place while providing a secure and controlled way to access them.

## Why Kubernetes secrets are not enough?

Kubernetes secrets are a great way to store sensitive information in a Kubernetes cluster. Kubernetes secrets are created and stored as a Kubernetes resource object and then used by pods or other resources inside the cluster.

Here is an example of a Kubernetes secret manifest file:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: mysecret
type: Opaque
data:
  username: YWRtaW4= # admin
  password: YWRtaW4= # admin
```

or using `kubectl`:

```bash
kubectl create secret generic mysecret --from-literal=username=admin --from-literal=password=admin
```

> But [Kubernetes secrets](https://kubernetes.io/docs/concepts/configuration/secret/) have some limitations:

Reviewer: ```suggestion
But [Kubernetes secrets](https://kubernetes.io/docs/concepts/configuration/secret/) have several limitations. Here are the top ones: 
```

> But [Kubernetes secrets](https://kubernetes.io/docs/concepts/configuration/secret/) several limitations. Here are the top ones:

Reviewer 2: Kubernetes secrets **have** several limitations.

> - They are stored by default in etcd in base64 encoded format, which is not secure.

Reviewer 2: wrap etcd in backticks.

> - They are created either manually by using `kubectl` commands or by using a manifest file, making them hart to manage at scale.

Reviewer 2: spelling: "hart"->"hard"

Author: 🙈 🇩🇪

- Hard to manage and synchronize secrets across different environments and clusters.
- There is no default way to rotate secrets automatically.

> I am sure there are some more limitations, but these are the most common ones that I have encountered. This is where ESO shines by providing a way to manage secrets in a more secure and efficient way:

Reviewer: ```suggestion
ESO shines by tackling all of the challenges listed and providing a way to manage secrets in a more secure and efficient manner. Benefits include: 
```

- Mitigate security risks as secrets are stored in a secure external secret management system like Pulumi ESC, which provides more robust security features like encryption, access control, and auditing.

- Eliminate manual management of secrets by synchronizing them from external secret management systems into the Kubernetes cluster, reducing the need for manual intervention and minimizing the risk of human error.

- ESO provides a unified interface, making it easy to manage secrets across clusters and multi-cloud environments.

- ESO integrates with external secret management systems which ensure that access to secrets is tracked and an audit trail is maintained. This is crucial for security compliance and adhering to regulatory standards.

> - Can automate secrets rotation when the external secret management system has support it.

Reviewer: ```suggestion
- Automate secrets rotation when the external secret management system has support for it.
```

## External Secrets Operator architecture

The central part of the [ESO architecture](https://external-secrets.io/latest/introduction/overview/) is the extension of the Kubernetes API with a new [Custom Resource Definition (CRD)](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/), which defines where the secrets are stored and how they are synchronized into Kubernetes secrets. The ESO controller then retrieves the secrets from the external secret management system and creates or updates the Kubernetes secrets. Every time a secret is updated in the external secret management system, the ESO controller will synchronize the secret in the Kubernetes cluster and update the secret accordingly.

![ESO Architecture](arch1.png)

> ## Core resources of ESO at a Glance

Reviewer: ```suggestion
### Core resources of ESO
```

- `PushSecret`: With the PushSecret resource, you can push secrets from the Kubernetes cluster to the external secret management system. The Pulumi ESC provider for ESO supports this feature, which allows you to push secrets from the cluster into Pulumi ESC. This feature is useful when you create secrets through other means and want to synchronize them into Pulumi ESC to manage them centrally.

## How to Use Pulumi ESC and ESO Together

To demonstrate how to use Pulumi ESC and ESO together, we will store some secrets and configuration values in Pulumi ESC, then synchronize them into a Kubernetes cluster using ESO.

To follow along, you'll need the following prerequisites:

- A Kubernetes cluster (I will be using a local [KinD](https://kind.sigs.k8s.io/) cluster, but you can use any
  Kubernetes cluster)
- [Pulumi CLI](/docs/iac/download-install/) installed
- `kubectl` [CLI installed](https://kubernetes.io/docs/tasks/tools/) for some debugging

> 

Reviewer: Add a note like that you can do the same using the ESC CLI

```bash
This command will walk you through creating a new Pulumi project.

Enter a value or leave blank to accept the (default), and press <ENTER>.
Press ^C at any time to quit.

Project description (A minimal Kubernetes TypeScript Pulumi program):
Created project 'eso-pulumi-esc'

Please enter your desired stack name.
To create a stack in an organization, use the format <org-name>/<stack-name> (e.g. `acmecorp/dev`).
Stack name (dev):
Created stack 'dev'

The package manager to use for installing dependencies npm
Installing dependencies...

added 310 packages, and audited 311 packages in 12s

44 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
Finished installing dependencies

Your new project is ready to go! ✨

To perform an initial deployment, run `pulumi up`
```

Head to Pulumi Cloud Console and create a new Pulumi ESC project with the name `eso-esc-demo`:

![Pulumi ESC Environment](env1.png)

And the environment `eso-esc-dev`:

![Pulumi ESC Environment](env2.png)

In the editor add the following yaml into the `Environment definition`:

```yaml
values:
  pulumiConfig:
    pulumi-pat:
      fn::secret: <your-pulumi-pat>
```

> {{% notes type="info" %}}

Reviewer: I could expand this to actually say "Pulumi access token" and not keep it as a note

```bash
curl -s  localhost:9898/env | grep FROM
  "FROM_ESC_VIA_ESO=world",
  "FROM_ESC_VIA_ESO_2=world",
```

And here we see secret values from the ESC project, proxied by ESO, and passed to the PodInfo application as environment variables. That means the secret was successfully mounted into the pod and the application is running as expected.

### Step 4: Clean Up

After you are done with the demo, you can clean up the resources by running:

```bash
pulumi destroy
```

## Conclusion

The combination of Pulumi ESC and External Secrets Operator provides a secure and efficient way to manage secrets in a cloud-native environment, improving security, efficiency, and compliance when consuming secrets in your Kubernetes cluster. By using Pulumi ESC, you can manage secrets and configurations in a centralized way, and with ESO, you now have a streamlined process to synchronize secrets into your Kubernetes cluster.

Start using Pulumi ESC and ESO today by creating an account on the Pulumi Cloud Console and begin managing your secrets in a secure and efficient way.

> 

Reviewer: Add something like "Don't have a Pulumi account yet. Click here to create an account and starting using Pulumi ESC"

---
title: "Pulumi ESC and External Secrets Operator: The Perfect Solution for Today's Cloud-Native Secret Management"
allow_long_title: true
date: "2024-10-01"
meta_desc: "Learn how to manage Kubernetes secrets with Pulumi ESC and External Secrets Operator"
meta_image: meta.png
authors:
- engin-diri

tags:
- pulumi
- esc
- external-secrets-operator
- kubernetes
- cloud-native
- security
- secret-management
---

> Managing secrets in a cloud-native environment can be challenging to do the right way. But it is crucial for ensuring

Reviewer 3: There might be a way to combine these first two sentences if ya want? i.e. Securing secrets in a cloud-native environment can be quite a challenge, but it's absolutely vital for safeguarding the security and integrity of any application or infrastructure.

the security and integrity of any application or infrastructure. We encounter a lot of different types of secrets, from
API keys, database passwords, and certificates to tokens and passwords. These secrets need to be stored securely and
accessed by different services in a secure way without exposing any sensitive information to unauthorized users.

Here is where Pulumi ESC and External Secrets Operator come into play by providing a secure and efficient solution for
cloud-native secret management.

## Pulumi ESC (Environments, Secrets, and Configuration)

Pulumi ESC is our new secrets management & orchestration service for environments, secrets, and configurations. It can
be used as part of your Pulumi IaC (Infrastructure as Code) projects using `pulumiConfig` to expose the stored values to
your Pulumi stacks.

But it can also be used as a standalone service to manage secrets and configurations. Pulumi ESC provides support for

> applications through its dedicated SDKs built for different programming languages. It also provides a dedicated CLI

Reviewer 3: ```suggestion
applications through its dedicated SDKs built for different programming languages. It also provides a dedicated CLI,
```

which can be used to manage secrets and configurations from the command line to provide a context for your CLI tools (
e.g. `terraform`).

For more information on Pulumi ESC, you can check out the official [documentation](/docs/esc/).

## External Secrets Operator (ESO)

The Open Source project [External Secrets Operator](https://external-secrets.io/latest/) (ESO) is a Kubernetes operator
that helps manage secrets in a highly secure and scalable way. ESO synchronizes secrets stored in external secret
management systems like Pulumi ESC, HashiCorp Vault, AWS Secrets Manager, Azure Key Vault, and more into Kubernetes
secrets. This approach allows us to control secret sprawl and centralize the management of secrets in a single place
while providing a secure and controlled way to access them.

> ## Why Kubernetes Secrets Are Not Enough?

Reviewer 3: ```suggestion
## Why Kubernetes secrets are not enough?
```

Kubernetes secrets are a great way to store sensitive information in a Kubernetes cluster. Kubernetes secrets are

> created and stored as a Kubernetes resource object as then used by pods or other resources inside the cluster.

Reviewer 3: ```suggestion
created and stored as a Kubernetes resource object and then used by pods or other resources inside the cluster.
```

Here is an example of a Kubernetes secret manifest file:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: mysecret
type: Opaque
data:
  username: YWRtaW4= # admin
  password: YWRtaW4= # admin
```

or using `kubectl`:

```bash
kubectl create secret generic mysecret --from-literal=username=admin --from-literal=password=admin
```

But [Kubernetes secrets](https://kubernetes.io/docs/concepts/configuration/secret/) have some limitations:

> - They are stored in etcd in base64 encoded format, which is not secure.

Author: Yes i know, I wanted to build the narrative 🙈

- They are created either manually by using `kubectl` commands or by using a manifest file, making them hart to manage
  at scale.
- Hard to manage and synchronize secrets across different environments and clusters.
- There is no default way to rotate secrets automatically.

I am sure there are some more limitations, but these are the most common ones that I have encountered. This is where ESO
shines by providing a way to manage secrets in a more secure and efficient way:

- Mitigate security risks as secrets are stored in a secure external secret management system like Pulumi ESC, which
  provide more robust security features like encryption, access control, and auditing.

- Eliminate manual management of secrets by synchronizing secrets from external secret management systems into the
  Kubernetes cluster. This eliminates the need to create and manage secrets manually and reduces the risk of human
  error.

- Breeze to manage cross-cluster and multi-cloud environments by providing unified interfaces to manage secrets across
  different environments and clusters.

> - In track with auditing and compliance are fulfilled as ESO integrates with external secret management and ensures

Reviewer 3: Maybe more like "Ensure auditing and compliance requirements are fulfilled ..."

  that access to secrets is tracked and an audit trail is maintained. This is crucial for compliance with security and

>   demonstrate the adherence to regulatory standards.

Reviewer 3: ```suggestion
  to demonstrate the adherence to regulatory standards.
```

- Automate secret rotation by providing a way to rotate secrets automatically by using the external secret management
  that supports secret rotation.

> ## External Secrets Operator Architecture

Reviewer 3: ```suggestion
## External Secrets Operator architecture
```

> Central part of the [ESO architecture](https://external-secrets.io/latest/introduction/overview/) is the extension of

Reviewer 3: ```suggestion
The central part of the [ESO architecture](https://external-secrets.io/latest/introduction/overview/) is the extension of
```

the Kubernetes API with a
new [Custom Resource Definition (CRD)](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/)

> which defines where the secrets are stored and how they are synchronized into Kubernetes secrets. The ESO controller

Reviewer 3: ```suggestion
, which defines where the secrets are stored and how they are synchronized into Kubernetes secrets. The ESO controller
```

then retrieves the secrets from the external secret management system and creates or updates the Kubernetes secrets.
Every time a secret is updated in the external secret management system, the ESO controller will synchronize the secret
in the Kubernetes cluster and update the secret accordingly.

![ESO Architecture](arch1.png)

> ## Core Resources of ESO in a glance:

Reviewer 3: ```suggestion
## Core resources of ESO at a glance
```

- `(Cluster)SecretStore`: Defines the external secret management system where the secrets are stored. It contains the

>   necessary information to connect to the external secret management system like the endpoint, authentication

Reviewer 3: ```suggestion
  necessary information to connect to the external secret management system, such as the endpoint, authentication
```

  credentials, and other configuration options. The `SecretStore` is namespaced while the `ClusterSecretStore` is
  cluster wide, which allows you cross-namespaced access to the secrets.

- `(Cluster)ExternalSecret`: Defines the secret that needs to be synchronized into the Kubernetes cluster. It contains
  the reference to the `SecretStore` and the secret key that needs to be synchronized and how the data should be
  transformed and saved as a Kubernetes secret.

- `PushSecret`: With the PushSecret resource, you can push secrets from the Kubernetes cluster to the external secret
  management system. The Pulumi ESC provider for ESO supports this feature, which allows you to push secrets from the

>   cluster into Pulumi ESC. This feature is useful when you created secrets through other means and want to synchronize

Reviewer 3: ```suggestion
  cluster into Pulumi ESC. This feature is useful when you create secrets through other means and want to synchronize
```

  them into Pulumi ESC to manage them centrally.

## How to Use Pulumi ESC and ESO Together

To demonstrate how to use Pulumi ESC and ESO together, we will create a simple example where we will store different

> secretes and configurations in Pulumi ESC and synchronize them into a Kubernetes cluster using ESO.

Reviewer 3: ```suggestion
secrets and configurations in Pulumi ESC and synchronize them into a Kubernetes cluster using ESO.
```

To follow along, you need to have the following prerequisites:

> - A Kubernetes cluster (I will be using a local kind cluster, but you can use any Kubernetes cluster)

Reviewer 2: "...local kind cluster..."->"...local cluster..."

> - Pulumi CLI installed

Reviewer 2: Add link to IaC CLI install docs

> - `kubectl` CLI installed for some debugging

Reviewer 2: Link to `kubectl` install docs

### Step 1: Install ESO and Link to Pulumi ESC

> We will use Pulumi with the `pulumi-kuernetes` provider to deploy ESO into the Kubernetes cluster. As we need to provide

Reviewer 2: typo: `pulumi-kuernetes`->`pulumi-kubernetes`

a PAT (Personal Access Token) to authenticate with Pulumi ESC, we will create a secret in the Kubernetes cluster to

> store. But this secret will be retrieved from Pulumi ESC using the Pulumi ESC integration of Pulumi IaC.

Reviewer 2: Might need to explain this more clearly... something like:

"We will be using ESC in two ways: First, to supply the Pulumi access token to ESO, via Pulumi IaC, which will allow the new ESO instance to access ESC later. Second, to provide configuration and secrets from ESC, via ESO, to the PodInfo application within the Kubernetes cluster."

{{% notes type="info" %}}
This is Pulumi ESC inception in action!
{{% /notes %}}

```bash
pulumi new kubernetes-typescript --name eso-pulumi-esc
```

You should see the following output:

```bash
This command will walk you through creating a new Pulumi project.

Enter a value or leave blank to accept the (default), and press <ENTER>.
Press ^C at any time to quit.

Project description (A minimal Kubernetes TypeScript Pulumi program):
Created project 'eso-pulumi-esc'

Please enter your desired stack name.
To create a stack in an organization, use the format <org-name>/<stack-name> (e.g. `acmecorp/dev`).
Stack name (dev):
Created stack 'dev'

The package manager to use for installing dependencies npm
Installing dependencies...

added 310 packages, and audited 311 packages in 12s

44 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
Finished installing dependencies

Your new project is ready to go! ✨

To perform an initial deployment, run `pulumi up`
```

> Head to Pulumi Cloud Console and create a new Pulumi ESC project with the name `eso-esc-demo:

Reviewer 2: typo: ``` `eso-esc-demo ```->``` `eso-esc-demo` ```

![Pulumi ESC Environment](env1.png)

And the environment `eso-esc-dev`:

![Pulumi ESC Environment](env2.png)

In the editor add the following yaml into the `Environment definition`:

```yaml
values:
  pulumiConfig:
    pulumi-pat:
      fn::secret: <your-pulumi-pat>
```

{{% notes type="info" %}}

> You don't know how to create a PAT? Check out the official

Reviewer 2: strike "You"

Pulumi [documentation](https://www.pulumi.com/docs/pulumi-cloud/access-management/access-tokens/).

{{% /notes %}}

Now, we need to link the Pulumi ESC project to the Pulumi IaC project. To do this, we need to add the following to your
`Pulumi.dev.yaml`:

```yaml
environment:
- eso-esc-demo/eso-esc-dev
```

> After all wiring is done, we can deploy the ESO into the Kubernetes cluster. Open the `index.ts` file and add the

Reviewer 2: "After all wiring is done..."->"Now that all the wiring is done..."

Reviewer 2: "...the ESO into..."->"...the ESO instance into..."

following code:

```typescript
import * as k8s from "@pulumi/kubernetes";
import * as pulumi from "@pulumi/pulumi";

const config = new pulumi.Config();

// Create a Kubernetes namespace
const ns = new k8s.core.v1.Namespace("external-secrets", {
    metadata: {
        name: "external-secrets",
    },
});

// Deploy a Helm release into the namespace
const externalSecrets = new k8s.helm.v3.Release("external-secrets", {
    chart: "external-secrets",
    version: "0.10.4", // Specify the version of the chart
    namespace: ns.metadata.name,
    repositoryOpts: {
        repo: "https://charts.external-secrets.io",
    },
});

// Deploy a secret into the namespace
const mySecret = new k8s.core.v1.Secret("my-secret", {
    metadata: {
        namespace: ns.metadata.name,
        name: "pulumi-access-token",
    },
    stringData: {
        "PULUMI_ACCESS_TOKEN": config.require("pulumi-pat"),
    },
    type: "Opaque",
});

// Export the namespace name
export const namespaceName = ns.metadata.name;

// Export the Helm release name
export const helmReleaseName = externalSecrets.name;
```

Deploy the stack by running:

```bash
pulumi up
```

> And you should see that the secret was created in the Kubernetes cluster and the ESO was deployed successfully.

Reviewer 2: "... the ESO was..."->"...the ESO instance was..."

```bash
kubectl get secret -n external-secrets pulumi-access-token -o jsonpath='{.data.PULUMI_ACCESS_TOKEN}' | base64 -d
```

### Step 2: Create an External Secret

Now, we will create an external secret in the Pulumi ESC project and synchronize it into the Kubernetes cluster. To do

> this, we need to create a new secret in the Pulumi ESC project and reference it in the ESO.

Reviewer 2: Strike redundant second sentence. eg.

"Now, we will create a secret in the Pulumi ESC project and synchronize it into the Kubernetes cluster by creating an ESO reference to the external secret."

> Create a new environment called `esp-esc-app` in the Pulumi ESC project with the name `eso-esc-demo`:

Reviewer 2: ...->"Create a new ESC environment called `esp-esc-app` in the `eso-esc-demo` ESC project."

![Pulumi ESC Environment](env3.png)

> And add the following yaml into the `Environment definition`:

Reviewer 2: ...->"Add the following YAML into the environment definition:"

```yaml
values:
  app:
    hello: world
    hello-secret:
      fn::secret: world
```

![Pulumi ESC Environment](esc.png)

> We can now create in the `index.ts` our `ClusterExternalSecret` and reference the secret from the Pulumi ESC project:

Reviewer 2: word order: "... create in the `index.ts` our `ClusterExternalSecret` and ..."->"... create the `ClusterExternalSecret` in `index.ts` and ..."

Reviewer 2: simplify: "...reference the secret from the Pulumi ESC project"->"...reference the ESC secret."

```typescript
import * as k8s from "@pulumi/kubernetes";
import * as pulumi from "@pulumi/pulumi";

// cut for brevity

const crd = new k8s.apiextensions.CustomResource("cluster-secret-store", {
    apiVersion: "external-secrets.io/v1beta1",
    kind: "ClusterSecretStore",
    metadata: {
        name: "secret-store",
    },
    spec: {
        provider: {
            pulumi: {
                organization: "dirien",
                project: "eso-esc-demo",
                environment: "esp-esc-app",
                accessToken: {
                    secretRef: {
                        name: mySecret.metadata.name,
                        key: "PULUMI_ACCESS_TOKEN",
                        namespace: mySecret.metadata.namespace,
                    },
                },
            },
        },
    },
});
```

> We can check that the secret store was successfully synchronized by running:

Reviewer 2: simplify: "...the secret store was..."->"...the secret was..."

```bash
kubectl get clustersecretstore secret-store -o jsonpath='{.status.conditions[?(@.type=="Ready")].status}'
```

### Step 3: Deploy an Application and Reference the Secret

Now, we can deploy an application that references the secret from the Kubernetes cluster. I am going to deploy the app

> called `PodInfo` which is a simple application that exposes the pod information as an API. Add the following code to the

Reviewer 3: ```suggestion
called `PodInfo`, which is a simple application that exposes the pod information as an API. Add the following code to the
```

> called `PodInfo` which is a simple application that exposes the pod information as an API. Add the following code to the

Reviewer 2: Strike "simple"

                name: "FROM_ESC_VIA_ESO_2",
                valueFrom: {
                    secretKeyRef: {
                        name: "esc-secret-store",
                        key: "hello",
                    }
                }
            }
        ]
    },
});

const externalSecretPodInfo = new k8s.apiextensions.CustomResource("external-secret-podinfo", {
    apiVersion: "external-secrets.io/v1beta1",
    kind: "ExternalSecret",
    metadata: {
        name: "esc-secret-store",
        namespace: podInfo.namespace,
    },
    spec: {
        dataFrom: [
            {
                extract: {
                    conversionStrategy: "Default",
                    key: "app",
                }
            }
        ],
        refreshInterval: "10s",
        secretStoreRef: {
            kind: clusterSecretStore.kind,
            name: clusterSecretStore.metadata.name,
        }
    },
});
```

This will deploy the `PodInfo` application and create an `ExternalSecret` resource that references the secret from the
Pulumi ESC project. We mount the secret into the pod as an environment variable by using the `extraEnvs` field in the
Helm chart.

{{% notes type="info" %}}

> In your real-world application, you could add different ways to reference the secret, like mounting it as a volume or

Reviewer 2: ...->"In a real-world application, you might use alternate ways to reference the secret, like mounting it as a volume..."

using `fromEnv` field in the Helm chart. It depends on your application and how you want to consume the secret.

{{% /notes %}}

> After deploying the stack, you should see that the `PodInfo` application is running and if you port-forward the service

Reviewer 2: run on sentence: ...->"After deploying the stack, you should see that the `PodInfo` application is running. To access the application, we can port-forward the service by running:"

by running:

```bash
kubectl port-forward svc/podinfo 9898:9898 -n podinfo
```

> You can access the API by running:

Reviewer 2: Need more explanation here of the final step. Something like:

"The `PodInfo` application provides an API that exposes information about the pod's environment. For example, we can read the environment variables set on the application process by accessing its `/env` endpoint. Let's try that:"

```bash
curl -s  localhost:9898/env | grep FROM
  "FROM_ESC_VIA_ESO=world",
  "FROM_ESC_VIA_ESO_2=world",
```

> And you should see that the secret was successfully mounted into the pod and the application is running as expected.

Reviewer 3: ```suggestion
Now you should see that the secret was successfully mounted into the pod and the application is running as expected.
```

> And you should see that the secret was successfully mounted into the pod and the application is running as expected.

Reviewer 2: ...->"And here we see secret values from the ESC project, proxied by ESO, and passed to the `PodInfo` application as environment variables. That means the secret was successfully mounted into the pod and the application is running as expected."

### Step 4: Clean Up

After you are done with the demo, you can clean up the resources by running:

```bash
pulumi destroy
```

## Conclusion

The combination of Pulumi ESC and External Secrets Operator provides a secure and efficient way to manage secrets in a

> cloud-native environment, guaranteeing security, efficiency, and compliance when consuming secrets in your Kubernetes

Reviewer 2: "guaranteeing"->"improving"

> cluster. By using Pulumi ESC, you can manage secrets and configurations in a centralized way, and with ESO, have a

Reviewer 3: ```suggestion
cluster. By using Pulumi ESC, you can manage secrets and configurations in a centralized way, and with ESO, you now have a
```
